### PR TITLE
Weighted-Cost Multi-Path HLD

### PR DESCRIPTION
**Pull request by nazariig**
Signed-off-by: Nazarii Hnydyn <nazariig@nvidia.com>

This document provides general information about Weighted-Cost Multi-Path implementation in SONiC

**In scope:**  
1. WCMP L3 configuration

**Out of scope:**  
1. WCMP EVPN Type-5 configuration

| PR title                                                           | state                                                                                                                  | context                                                                                                                     |
| ------------------------------------------------------------------ | ---------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------- |
| [BGP OA](https://github.com/sonic-net/sonic-buildimage/pull/18533) | GitHub issue/pull request detail | GitHub pull request check contexts |
| [CLI](https://github.com/sonic-net/sonic-utilities/pull/3253)      | GitHub issue/pull request detail   | GitHub pull request check contexts   |

### CODE REVIEW

File: doc/wcmp/wcmp-design.md [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+uses the same next hop, and by extension, the same path across the network.
+
+Weighted ECMP using BGP link bandwidth introduces support for network-wide UCMP to an IP destination.  
+The unequal cost load balancing is implemented by the forwarding plane based on the weights associated  
+with the next hops of the IP prefix. These weights are computed based on the bandwidths of the corresponding  
+multipaths which are encoded in the BGP link bandwidth extended community. Exchange of an appropriate  
+BGP link bandwidth value for a prefix across the network results in network-wide unequal cost multipathing.
+
+WCMP is applicable in a pure L3 network as well as in a EVPN network.
+
```
Comment (kperumalbfn) [inline doc/wcmp/wcmp-design.md:129]: Sonic WCMP HLD that was discussed earlier - https://github.com/sonic-net/SONiC/commit/078eb697adb58f5e8ae1dabeedd87419069bd91f

HLD was not merged, but the swss changes are merged. https://github.com/sonic-net/sonic-swss/pull/1853

Are we using this implementation?

**Comment by zhangyanzhao** [comment]
community review recording https://zoom.us/rec/share/Wlvm6_mZtV5Ov-QihEY_bt3JxVFlizC7jRMIwVQKwNJy4s2Ub5sXuRzAV0dyj1Aa.3uxRviw8exKORObs

File: doc/wcmp/wcmp-design.md (hunk 1 above)
Comment (nazariig) [reply doc/wcmp/wcmp-design.md:129]: @kperumalbfn yes

File: doc/wcmp/wcmp-design.md [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+Weighted-Equal Cost-Multipath (W-ECMP) offers a solution to this problem in which the relative link weight  
+reflecting the available capacity is distributed using the BGP bandwidth community attribute.
+
+Dataplane configuration flow:
+* The BGP link bandwidth extended community carries information about the available link capacity for the route  
+prefixes through the network, which gets mapped to the weight of the corresponding next hop. The mapping  
+factors the bandwidth value of a particular path against the total bandwidth values of all possible paths,  
+normalized to the range 1 to 255
+* FRR converts the incoming BGP link bandwidth extended community values into proportionated weight  
+among the ecmp members in such a way that the cumulative value of individual weights is normalized to 255
```
Comment (philo-micas) [inline doc/wcmp/wcmp-design.md:218]: AS individual weight is normalized to be between 1 and 255, what does  "the cumulative value of individual weights  is normalized to 255" mean?

Comment (nazariig) [reply doc/wcmp/wcmp-design.md:218]: @philo-micas the answer to this question will be added later to the official documentation here: [weighted-ecmp-using-bgp-link-bandwidt](https://docs.frrouting.org/en/latest/bgp.html#weighted-ecmp-using-bgp-link-bandwidth)
The relevant request was forwarded to the FRR team.

**Comment by liat-grozovik** [comment]
@kperumalbfn  @philo-micas  any further comments or you will be able to approve the HLD?
Code PRs are avaialble, do you wish to take part of the PRs review as well?

@zhangyanzhao  need to track it for 202405. This feature need to get in.

**Review by stepanblyschak — APPROVED** [review]

**Comment by zhangyanzhao** [comment]
Add @tjchadaga as reviewer and please help to review this PR.

File: doc/wcmp/wcmp-design.md [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+    advertise ipv4 unicast route-map wcmp-map
+    neighbor SPINE activate
+  exit-address-family
+  !
+end
+```
+
+**Note:** EVPN configuration is only relevant when `docker_routing_config_mode` is either `split` or `split-unified`
+
+## 2.3 Orchestration agent
```
Comment (venkatmahalingam) [inline doc/wcmp/wcmp-design.md:294]: We need to rename this section as this section talks about programming the WCMP configs into the FRR via bgpcfgd.

You can link the existing the WCMP HLD created with orchestration agent changes here.

File: doc/wcmp/wcmp-design.md [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+
+No special handling is required
+
+# 3 Test plan
+
+## 3.1 Unit tests via PYTEST
+
+WCMP basic configuration test:
+1. Verify FRR route-map is generated properly after WCMP config update
+
```
Comment (venkatmahalingam) [inline doc/wcmp/wcmp-design.md:598]: We may need a UT to verify whether the WCMP configs are applied after save and reboot.

Comment (nazariig) [reply doc/wcmp/wcmp-design.md:598]: @venkatmahalingam simple UTs are not intended for such a testing

Comment (venkatmahalingam) [reply doc/wcmp/wcmp-design.md:598]: Ok, can you point to the comprehensive list of test cases for this feature?

Comment (nazariig) [reply doc/wcmp/wcmp-design.md:598]: @venkatmahalingam sure. I will update you once the test plan is ready

File: doc/wcmp/wcmp-design.md (hunk 3 above)
Comment (nazariig) [reply doc/wcmp/wcmp-design.md]: @venkatmahalingam done

**Comment by liat-grozovik** [comment]
@philo-micas @venkatmahalingam kindly reminder to review and approve if all comments were handled

**Comment by lguohan** [comment]
@philo-micas , @venkatmahalingam , what is pending on this HLD?

**Review by venkatmahalingam — APPROVED** [review]

**Review by liat-grozovik — APPROVED** [review]

**Comment by zhangyanzhao** [comment]
Nvidia will link the test PR and BRCM will help to review the code PRs. @adyeung  @liat-grozovik

**Comment by nazariig** [comment]
> Nvidia will link the test PR and BRCM will help to review the code PRs. @adyeung @liat-grozovik

@zhangyanzhao done

**Comment by zhangyanzhao** [comment]
code PRs are still open, move to backlog for future release

**Pull request merged**